File: solutions/020.valid-parentheses/valid-parentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = []
        pair = {')': '(', ']': '[', '}': '{'}
        for p in s:
            if p in ['(', '[', '{']:
                stack.append(p)
            else:
                if not stack  or stack.pop() != pair.get(p):
                    return False
        if stack:
            return False
        return True
            
        
        

File: solutions/020.valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        char[] stack = new char[s.length() + 1];
        int top = 1;  // top is null
        for(int i=0; i < s.length(); i++){
            char c = s.charAt(i);
            if(c == '(' || c == '[' || c == '{') {
                stack[top++] = c;
            } else if(c == ')' && stack[--top] != '('){
                return false;
            }else if(c == ']' && stack[--top] != '['){
                return false;
            }else if(c == '}' && stack[--top] != '{'){
                return false;
            }
        }
        return top == 1;
    }
}